src/dev/lupio/lupio_blk.cc

```diff
@@ -0,0 +1,233 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_blk.hh"
+
+#include "debug/LupioBLK.hh"
+#include "dev/dma_device.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioBLK.hh"
+
+/* Shared fields for CTRL and STAT registers */
+#define LUPIO_BLK_TYPE  0x2
+
+/* Specific fields for CTRL */
+#define LUPIO_BLK_IRQE  0x1
+
+/* Specific fields for STAT */
+#define LUPIO_BLK_BUSY  0x1
+#define LUPIO_BLK_ERRR  0x4
+
+/* Define Int ID for PCI calls */
+#define LUPIO_BLK_INT_ID   0xb
+
+#define SECTOR_BITS   9
+#define SECTOR_SIZE   (1ULL << SECTOR_BITS)
+
+namespace gem5
+{
+LupioBLK::LupioBLK(const Params &params) :
+    DmaDevice(params),
+    platform(*params.platform),
+    dmaEvent([this]{ dmaEventDone(); }, name()),
+    pioAddr(params.pio_addr),
+    pioSize(params.pio_size),
+    image(*params.image)
+{
+    nbBlocks = image.size();
+    DPRINTF(LupioBLK, "LupioBLK initalized\n");
+}
+
+static inline int clz64(uint64_t val)
+{
+    return val ? __builtin_clzll(val) : 64;
+}
+
+void
+LupioBLK::dmaEventDone()
+{
+    if (writeOp) {
+        for (int i = 0; i < reqLen; i += SECTOR_SIZE){
+            image.write(&reqData[i], offset);
+            offset++;
+        }
+    }
+    delete[] reqData;
+    DPRINTF(LupioBLK, "Done with DMA event\n");
+    platform.postPciInt(LUPIO_BLK_INT_ID);
+}
+
+uint32_t
+LupioBLK::lupioBLKRead(const uint8_t addr)
+{
+    uint32_t r = 0;
+
+    switch (addr >> 2) {
+        case LUPIO_BLK_CONF:
+            r = SECTOR_BITS << 16               /* log2(512B / block) */
+                | (63 - clz64(nbBlocks));      /* log2(Nb of blocks) */
+
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_CONF: %d\n", r);
+            break;
+        case LUPIO_BLK_NBLK:
+            r = nblk;
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_NBLK: %d\n", r);
+            break;
+        case LUPIO_BLK_BLKA:
+            r = lba;
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_BLKA: %d\n", r);
+            break;
+        case LUPIO_BLK_MEMA:
+            r = mem;
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_MEMA: %d\n", r);
+            break;
+
+        case LUPIO_BLK_STAT:
+            r = 0;                      /* Always idle */
```

Comment:
It's always idle with QEMU because the block device has no latency but it's not true here. You should have a completion boolean or something that gets toggled in BLKCmd() and toggled back in DmaDone(). When the block device is performing a transfer, then the status register should accurately reflect that.